lib/llm/src/hub.rs

```diff
@@ -36,68 +32,120 @@ fn is_weight_file(filename: &str) -> bool {
 /// If ignore_weights is true, model weight files will be skipped
 pub async fn from_hf(name: impl AsRef<Path>, ignore_weights: bool) -> anyhow::Result<PathBuf> {
     let name = name.as_ref();
+    let model_name = name.display().to_string();
+    
+    // Get ModelExpress server endpoint from environment or use default
+    let endpoint = env::var(MODEL_EXPRESS_ENDPOINT_ENV_VAR)
+        .unwrap_or_else(|_| DEFAULT_MODEL_EXPRESS_ENDPOINT.to_string());
+    
+    // Create client configuration
+    let config = ClientConfig::for_testing(&endpoint);
+    
+    // First try to use the server with fallback
+    let result = match Client::new(config.clone()).await {
+        Ok(mut client) => {
+            // Try server-based download first
+            match client.request_model_with_provider_and_fallback(&model_name, ModelProvider::HuggingFace).await {
+                Ok(()) => {
+                    // Server download succeeded, now get the path
+                    get_model_path_from_cache(&model_name)
+                }
+                Err(e) => {
+                    // Server failed, fallback to direct download
+                    tracing::warn!("Server download failed for model '{}': {}. Falling back to direct download.", model_name, e);
+                    download_model_directly(&model_name).await
+                }
+            }
+        }
+        Err(e) => {
+            // Can't connect to server, use direct download
+            tracing::warn!("Cannot connect to ModelExpress server: {}. Using direct download.", e);
+            download_model_directly(&model_name).await
+        }
+    };
+    
+    // If ModelExpress methods fail, try hf-hub as final fallback
+    match result {
+        Ok(path) => Ok(path),
+        Err(e) => {
+            tracing::warn!("ModelExpress download failed for model '{}': {}. Falling back to hf-hub.", model_name, e);
+            download_with_hf_hub(&model_name, ignore_weights).await
+        }
+    }
+}
+
+/// Download model directly using ModelExpress download function
+async fn download_model_directly(model_name: &str) -> anyhow::Result<PathBuf> {
+    // Get cache directory
+    let cache_dir = get_model_express_cache_dir();
+    
+    // Use ModelExpress download function directly
+    download::download_model(model_name, ModelProvider::HuggingFace, Some(cache_dir)).await
+}
+
+/// Download model using hf-hub as final fallback
+async fn download_with_hf_hub(model_name: &str, ignore_weights: bool) -> anyhow::Result<PathBuf> {
+    tracing::info!("Downloading model '{}' using hf-hub", model_name);
+    
+    // Get cache directory
+    let cache_dir = get_model_express_cache_dir();
+    
+    // Get Hugging Face token if available
     let token = env::var(HF_TOKEN_ENV_VAR).ok();
+    
+    // Create hf-hub API client
     let api = ApiBuilder::new()
         .with_progress(true)
         .with_token(token)
         .high()
+        .with_cache_dir(cache_dir)
```

Comment:
Is the model express cache compatible with the HF cache? Is model express using hf-hub internally?

Reply:
I removed this, HF_HUB_CACHE should be the mount for the modelexpress client volume anyways

but there is no need to provide the path again to the hf_hub downloader